plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java

```diff
@@ -85,31 +91,39 @@ private Type getType(FieldDescriptor fieldDescriptor)
             case DOUBLE -> DOUBLE;
             case BYTE_STRING -> VARBINARY;
             case STRING, ENUM -> createUnboundedVarcharType();
-            case MESSAGE -> getTypeForMessage(fieldDescriptor);
+            case MESSAGE -> getTypeForMessage(fieldDescriptor, processedMessages);
         };
 
-        // Protobuf does not support adding repeated label for map type but schema registry incorrecty adds it
+        // Protobuf does not support adding repeated label for map type but schema registry incorrectly adds it
         if (fieldDescriptor.isRepeated() && !fieldDescriptor.isMapField()) {
             return new ArrayType(baseType);
         }
         return baseType;
     }
 
-    private Type getTypeForMessage(FieldDescriptor fieldDescriptor)
+    private Type getTypeForMessage(FieldDescriptor fieldDescriptor, PSet<String> processedMessages)
```

Comment:
Any specific reason for using `PSet` ?

Reply:
It's lightweight GC friendly implementation of chain of immutable Set(s). Otherwise it's need to be create the such object container or concat processed messages into some string and parse on every iteration. So I don't see any reason why not use existing well fitting implementation of `PSet` for that particular case with no adding nothing redundant.